File: math/313.py
# 从固定的数变成给出的primes数组, 需要用额外数组表示indexs跟values
import itertools
class Solution:
    def nthSuperUglyNumber(self, n, primes):
        """
        :type n: int
        :type primes: List[int]
        :rtype: int
        """
        if n<= 0:
            return 1
        ret = []
        ret.append(1)
        curvalue = list(itertools.repeat(0, len(primes)))
        indexs = list(itertools.repeat(0, len(primes)))
        
        while len(ret) < n:
            for i in range(len(primes)):
                curvalue[i] = ret[indexs[i]] * primes[i]
            temp = curvalue[0]
            for i in range(1, len(primes)):
                temp = min(temp, curvalue[i])
            for i in range(len(primes)):
                if temp == curvalue[i]:
                    indexs[i] += 1
            ret.append(temp)
        
        return ret.pop()
                
        

File: math/313.go
func nthSuperUglyNumber(n int, primes []int) int {
    dp := make([]int, n)
    dp[0] = 1
    indexs := make([]int, len(primes))
    for i := 1; i < n; i++ {
        dp[i] = math.MaxInt32
        for j := 0; j < len(primes); j++ {
            dp[i] = min(dp[i], dp[indexs[j]]*primes[j])
        }
        for j := 0; j < len(primes); j++ {
            if dp[i] == dp[indexs[j]]*primes[j] {
                indexs[j]++
            }
        }
    }
    return dp[n-1]
}

func min(i int, j int) int {
    if i < j {
        return i
    }else {
        return j
    }
}
